### ISSUE TITLE
User-assigned Managed Identity authorization issue 'Microsoft.Resources/subscriptions/resourcegroups/write

### ISSUE THREAD

**Issue by jlou07** [Azure/AVDBlueprint#127]
Hi Team,

After the ps1 was launched by PowerShell, my Blueprint was stopped by by this error message :

The resource group 'AVDBP01-sharedsvcs-rg' failed to deploy due to the following error: The client 'a6fb85be-bb6d-45e1-bbe0-186994ea1499' with object id 'a6fb85be-bb6d-45e1-bbe0-186994ea1499' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourcegroups/write' over scope '/subscriptions/26a351f5-e68a-40da-97dd-adf5b6e97f81/resourcegroups/AVDBP01-sharedsvcs-rg' or the scope is invalid. If access was recently granted, please refresh your credentials.

If I check the User-assigned Managed Identity created, I only see “Blueprint Operator” role :
![managed-identity](https://user-images.githubusercontent.com/58774660/135841653-8f39351e-5836-4887-9931-3536df2118cd.png)

Do you think some permissions are missing in that role in order to process the Blueprint ?

Jean-Loup

**Comment by robsmi-msfte** [comment]
Hello @jlou07,

Are you using the single assignment PowerShell script in 'Examples and Samples\Deploy with Single Script' folder?  If so, check the section that grants subscription ownership to the managed identity.  That section has output like this:

"Now checking if user assigned identity '$UserAssignedIdentityName' has 'Owner' subscription level role assignment"

What is the output here?  The managed identity has to have 'Owner' role at the subscription level, as well as 'Global Administrator' role in Azure AD.  These roles can be removed once the deployment is complete.

Thanks,

Robert Smith

**Comment by jlou07** [comment]
Hi Robert,

Thanks for your help and y
![error-roles](https://user-images.githubusercontent.com/58774660/135912660-84848db5-3e0e-451d-adb3-f67e2e4ae0df.png)
es I am using that one.

You were right, I have got some error messages for both roles :
Get-AzureADMSRoleAssignment is an older command, isn'it ?

Thanks,

Jean-Loup

**Comment by jlou07** [comment]
Finaly I uninstall all AzureAD module and let the script install it.

But I get another error about role assignment 
![error-assign](https://user-images.githubusercontent.com/58774660/135920429-a2f680d8-a7ac-4181-8dba-8886c4581edf.png)

Therefore I have to add manually owner and blueprint operator roles at the subscription level

**Comment by robsmi-msfte** [comment]
Hello @jlou07,

If you get a chance to test again, I added a few lines of code in that section, to check for UAI and if it's not there, wait until it is.  I put that code in a new branch for testing, here:

https://github.com/Azure/AVDBlueprint/tree/TestUAI1

If you get a chance, copy that version of the single assignment script to your local folders and try again and let us know how it goes.  If that fixes the issue, I will incorporate that into the main branch.

I think the AzureRM commands are the ones deprecated.  I checked the commands used in the script to make sure they weren't deprecated.

Thanks,

Robert Smith

**Comment by jlou07** [comment]
Hello Robert,

Sure, I will test it tomorrow from scratch and let you know.
My blueprint is still deploying so it smells good 👍 

You are right about AzureRM (mismatch).
After uninstalling, things look like better.

Jean-Loup

**Comment by robsmi-msfte** [comment]
Sounds good.  Thanks for the update.

Thanks,

Robert M. Smith
________________________________

**Comment by jlou07** [comment]
Hi Robert,

I replace my folders by https://github.com/Azure/AVDBlueprint/tree/TestUAI1

But I still get an error message here :
![error-assign](https://user-images.githubusercontent.com/58774660/136097839-b4cfaf40-8599-4b6c-b037-388b208537d6.png)

And I do not see anything here (even F5 is my best friend)
![Azure-roles](https://user-images.githubusercontent.com/58774660/136098008-659acc5c-9aa6-4e46-a1bc-ab1023569dda.png)

I have to add Azure Roles manually and they appear.

Jean-Loup

**Comment by robsmi-msfte** [comment]
Hello Jean-Loup,

I will have to give you contributor credit at this point.  Looking at the error below, it looks like a timing problem.  You create an object, then if you go to access it too quickly, you get an error back that the object doesn't exist.  The error GUID is the PrincipalID of the managed identity.  I have seen this before and have had to code pauses in to account for that.  What I did just now in the branch 'TestUAI1', I added a "Do Until" loop that checks for the existence of the managed identity, before attempting the next operation.  I did that like this:

Write-Host "Now checking if user assigned identity '$UserAssignedIdentityName' has 'Blueprint Operator' subscription level role assignment" -ForegroundColor Cyan
$UAMIBlueprintOperatorRoleCheck = Get-AzUserAssignedIdentity -Name $UserAssignedIdentityName -ResourceGroupName $BlueprintGlobalResourceGroupName
if (-not($UAMIBlueprintOperatorRoleCheck)) {
    Do {
    Write-Host "User assigned identity '$UserAssignedIdentityName' is not currently available, waiting 3 seconds..." -ForegroundColor Cyan
    Start-Sleep -Seconds 3
    } until (Get-AzUserAssignedIdentity -Name $UserAssignedIdentityName -ResourceGroupName $BlueprintGlobalResourceGroupName -ErrorAction SilentlyContinue)
    Write-Host "User Assigned Managed Identity '$UserAssignedIdentityName' is now available..." -ForegroundColor Cyan
}

I also moved the section that registers the Blueprint provider, up above this block.  I also encoded a pause in there, to wait for the Blueprint provider registration state to change from "registering" to "registered".  This is just a safeguard to ensure the provider is up and ready to go, before the script moves to the next thing.

I apologize that you have encountered these problems.  If you can test this new code in file 'AssignAVDBlueprint.ps1', I hope it gets everything resolved.

Thanks,

Robert Smith

**Comment by robsmi-msfte** [comment]
Closing this issue, because we have developed workarounds and/or resolutions.  Please open a new issue if we encounter further issues.
